### PR TITLE
#41 remove subscribers

### LINKED ISSUE #41: Create a remove_subscribers.py script
# Description
Develop a Python script that checks the RDS database to identify products being tracked. If a product no longer has any users subscribed to it, the script should remove the product's URL from the database and stop tracking it. The solution should also include tests for the script to ensure functionality and adhere to Test-Driven Development (TDD) practices.

# Outcome
- `remove_subscribers.py`: A Python script that connects to the RDS database to iterate over tracked products. It verifies if each product has any users subscribed to it. If no users are found, it removes all database rows associated with the product ID and stops tracking the product, unless a user subscribes to it again in the future. 
- `test_remove_subscribers.py`: A suite of tests for remove_subscribers.py to enforce correctness and follow TDD principles, ensuring expected behavior in edge cases.

# User story
As a systems administrator, I want a script to automatically clean up and stop tracking products with no active subscriptions, so that the database remains optimized, and unnecessary processing for inactive products is avoided.

### PR DESCRIPTION
**Pull request by gem09lo**
Added a script that cleans the rds database for any products that have no active subscriptions.
- It retrieves all of the `product_id`s from the subscription schema
- Retrieves all of the `product_id`'s of the current products being tracked in product schema
- Finds all of the unsubscribed products (based on the above two queries)
- Deletes the rows that have those product_ids without subscriptions from the price_changes and product schema (and the websites that are no longer tracked).

This partially closes #41 - just need to remove the functions that involve datetime

### CODE REVIEW

File: pipeline/remove_subscribers.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def clean_websites(cur) -> None:
+    """Removes websites not referenced in product table."""
+    query = """DELETE FROM website
+    WHERE website_id NOT IN (
+        SELECT DISTINCT website_id
+        FROM product);"""
+    cur.execute(query)
+    logging.info("Unused websites have been removed.")
```
Comment (jk-george) [inline pipeline/remove_subscribers.py:54]: These three functions, they're primarily just cursor.execute(). Since they're carrying out the same purpose it could be counted as "repeated code".

Comment (jk-george) [reply pipeline/remove_subscribers.py:54]: May also want to consider the same for the "get_all_products" and "get_active_subscriptions"

File: pipeline/remove_subscribers.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    """Deletes data for unsubscribed products from all relevant tables."""
+    try:
+        with conn.cursor() as cur:
+            delete_price_changes(cur, unsubscribed_product_ids)
+            delete_products(cur, unsubscribed_product_ids)
+            clean_websites(cur)
+        conn.commit()
+    except Exception as e:
+        conn.rollback()
+        logging.error("Error deleting unsubscribed data: %s", e)
```
Comment (jk-george) [inline pipeline/remove_subscribers.py:67]: Good use of logging.

**Review by jk-george — DISMISSED** [review]
Good work creating tests for the py file. Just a few things in here to consider.

File: pipeline/remove_subscribers.py [hunk 3]
```diff
@@ -1,50 +1,32 @@
...
-        return set()
-
-
-def get_all_products(conn: connection) -> set:
-    """Returns all product_ids from product table."""
-    try:
-        with conn.cursor() as cur:
-            cur.execute("SELECT product_id FROM product;")
+            add_distinct = "DISTINCT" if distinct else ""
+            query = f"""SELECT {add_distinct} product_id FROM {table_name};"""
```
Comment (jk-george) [inline pipeline/remove_subscribers.py:14]: Nice way of adding DISTINCT to the string.

File: pipeline/remove_subscribers.py [hunk 4]
```diff
@@ -1,50 +1,32 @@
...
 from psycopg2.extensions import connection
+from psycopg2.extras import DictCursor
 from dotenv import load_dotenv
 from connect_to_database import configure_logging, get_connection
 
 
-def get_active_subscriptions(conn: connection) -> set:
-    """Returns all product_ids that have users subscribed to it from subscription schema."""
+def get_product_ids_from_table(conn: connection, table_name: str, distinct: bool = False) -> set:
+    """Returns a set of product IDs based on specified schema."""
```
Comment (jk-george) [inline pipeline/remove_subscribers.py:10]: Very nice way to slim down the code, especially leaving a default for the distinct unless stated.

File: pipeline/remove_subscribers.py [hunk 5]
```diff
@@ -1,50 +1,32 @@
...
 
-def delete_price_changes(cur, unsubscribed_product_ids: list) -> None:
-    """Deletes price changes for unsubscribed products."""
-    query = """DELETE FROM price_changes WHERE product_id = ANY(%s);"""
-
-    cur.execute(query, (unsubscribed_product_ids,))
-    logging.info("Removed price changes for products: %s",
-                 unsubscribed_product_ids)
-
+def delete_from_table(cur: DictCursor, table_name: str, product_ids: list) -> None:
```
Comment (jk-george) [inline pipeline/remove_subscribers.py:22]: Similar to the above change, very nice!

**Review by jk-george — APPROVED** [review]
Looks really good! The changes are great.

File: pipeline/remove_subscribers.py (hunk 1 above)
Comment (jk-george) [reply pipeline/remove_subscribers.py]: This was fixed in the latest push to the branch.
